Specify GraphQL API for clients and nodes

adzialocha opened the pull request.

adzialocha commented:
I've added already some more "sophisticated" parameters for querying document views. We probably will not implement them for now, but I thought its cool to already get them specified (even if they might change slightly).

adzialocha commented on website/docs/01-writing-data/clients-nodes.md:
```diff
@@ -4,39 +4,52 @@ id: clients-nodes
-- nodes are computer programs that
-  - validate and store entries received from a client that connects to it
-  - exchange entries with other nodes
-  - process entries to make information in them available in the shape that clients need
```
Yeah, that would be great! Should we do it now or for the future as I can see that many other places would need diagrams? Also, I'm not sure if the diagrams should be on specification-level or on the "Learn"-section-level, or on both.

cafca replied:
Ah yeah, I didn't mean this PR, just a general observation.

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
+  """
+  flag indicating if this document has been deleted
+  """
+  deleted: Boolean!
```
I was thinking it would be cool to eventually make available more information for `edited` and `deleted`, i.e. make deleted an `Option<(PublicKey, OperationId>)` so we know who has deleted the document without making an additional request and similarly making `edited` a list of public keys who touched the document with their respective most recent edit operation id. 

I don't think we need this now but was asking myself if introducing that would be a breaking api change later unless we keep both this boolean form and some new list-based form.

adzialocha replied:
I like the type of it being a simple `Boolean`, otherwise I'd need to do stuff like `const isDeleted = result.deleted.length === 0`, it tells me something about _the general state_ of this document which feels nice.

The feature you're suggesting is cool though, we should definitely have it, maybe it could live in its own field?

adzialocha replied:
Just that timestamps in this world do not exist 😅

cafca replied:
operation ids are our timestamps :)

cafca commented on website/docs/03-networking/replication.md:
```diff
@@ -4,11 +4,216 @@ id: replication
-
-- Replication is the process by which nodes exchange data
-    - p2panda uses _gossip replication_: nodes autonomously share data with each other without central coordination
+- replication is the process by which nodes exchange data to eventually converge all to the same state
```
Is data=entries, operations or is there also other data? Otherwise we could just say entries and operations here.

cafca commented on website/docs/03-networking/replication.md:
```diff
@@ -4,11 +4,216 @@ id: replication
+): SingleEntryAndPayload
+
+"""
+get aliases of the provided `public_keys` that you can use in future requests
+to save bandwidth
```
I feel like aliases could be expanded a bit later on, I still have some questions :D Like: When do aliases expire? Are they per node or per node-client-connection? How does the node tell me that it forgot about an alias but may still know the public key it's supposed to connect to?

cafca reviewed (changes requested):
I had some questions :)

sandreae commented on website/docs/03-networking/replication.md:
```diff
@@ -4,11 +4,216 @@ id: replication
+- nodes may not be interested in all available data and can choose to receive only some data, for example by only requesting documents following a certain schema
+- nodes need to implement the API specifications to make sure they are compatible with all other node and client implementations. The Node API is specified here, the Client API is further specified under [queries][queries], both APIs reside inside nodes
+
+## Node API
```
Oh, you did the replication spec too? Cool :+1:

adzialocha replied:
Yeah, at least the beginning. The defines the GraphQL API for nodes, but not the replication mechanism on top of that.

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
-- clients can retrieve materialised instances of a given schema
-    - _instances can be filtered by individual fields_
+    - before that, clients can retrieve parameters required for encoding entries if they can't compute them independently
+- clients can retrieve materialised documents and document views of a given schema
```
Do you always know the document id when you want to get a view id? Probably not right?

sandreae replied:
Yeh, that's what I was thinking too, I don't think you do (like in pinned relation fields).

We could of course return it in the document views' meta data, so from that point of view you could know.

adzialocha commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
+  <field_name>_ne: <value>
+
+  """
+  filter by fields containing larger values
```
I guess we can also do this for `String` if we specify how strings are sorted?

adzialocha replied:
Also `relation` and `pinned_relation` will work if they are self-referential

adzialocha replied:
.. aaaand `Float` is rather [tricky](https://floating-point-gui.de/errors/comparison/) actually :boom:

adzialocha replied:
Ah, Rust actually has a new workaround since `1.62`: https://doc.rust-lang.org/nightly/std/primitive.f64.html#method.total_cmp

adzialocha commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
+  hasNextPage: Boolean!
+
+  """
+  when paginating backwards, the cursor to continue.
```
The recommendation is to use an opaque identifier of the underlying data which is then base64 encoded (https://graphql.org/learn/pagination/#pagination-and-edges), so something rather "unrelated" to the underlying data, but unique. I guess in our case it can be a base64-encoded document view id, but as I read the [spec](https://relay.dev/graphql/connections.htm) it is up to the implementer to decide, as long as it fullfills its purpose to aid pagination.

We can add a note on that thought?

adzialocha replied:
I've added a note on that :+1:

adzialocha commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
+  first: Int
+
+  """
+  cursor identifier for pagination
```
Yes, similar to my comment somewhere else, its that opaque cursor type, which is probably a base64 encoded document view id.

adzialocha replied:
I've added a note on that :+1:

adzialocha commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
 
 - publishes the entry supplied with the request
-- returns entry arguments required for publishing the next entry for the same document
+  - the entry is validated by the receiving node and persisted in a database
```
That almost smells like its own chapter, but I can add some notes here

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -199,16 +203,35 @@ type <schema_id>ResponseFields {
 #### `all_<schema_id>`
 
-- returns many document views of a given schema
+- returns zero to many documents (the "latest" document views) of a given schema
```
If you feel the need to put "latest" in quotes maybe we should define that. Something like

> the _latest_ document view is defined as the view that incorporates all valid operations known to the node at this point in time. 

Not sure where the right place to put that would be though. Maybe not even here but in the document section?

adzialocha replied:
Hmn yeah, maybe I should remove that? It might just be misleading and one can read about it in the other specs.

cafca replied:
Yes, ideally linking to a section where one can learn more about this.

sandreae replied:
What about:

> returns the current view for many documents of a given schema

?

adzialocha commented:
Thank you for your great feedback @sandreae and @cafca! I've had another iteration on this and some things we're added and came up:

* Expanded on ordering / filtering on self-referential relations as they form a sortable, filterable list themselves, I've added a little bit more about it in the `reduce` section as well. Its not necessary for this PR, but I felt like doing it anyhow to be more precise about how we order operation ids before reducing them.
  * That brings up the question: Why only filter / sort self-referential relations? We _could_ have a topological ordering on _every_ relation graph, even when its refering to another schema? It might be computationally expensive though to do it for all relations, but its a cool feature?
* I've fixed the letter-casing of the GraphQL fields, guess we're doing lowerCamelCase, but how about the filters (`<field_name>_eq`) ..? I'd be fine with keeping it like this as it is a special thing
* I've tried to be more precise over the pagination by refering to the GraphQL spec and recommending an approach
* Sorting and comparing over "floats" is funny! I'm not sure how to deal with that as it only comes with a bunch of [caveats](https://floating-point-gui.de/) ..

sandreae reviewed (approved):
This reads great to me now :+1: Thanks!

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
 - returns parameters required for encoding new entries
     - no side effects
-- clients can't encode new entries without information from this endpoint because every entry needs to place itself in the first unused sequence number of a specific [_bamboo log_][encoding] and also it needs to include the hashes of specific previous entries in its encoding
+- clients can't encode new entries without information from this endpoint because every entry needs to place itself in the first unused sequence number of a specific [_bamboo log_][bamboo] and also it needs to include the hashes of specific previous entries in its encoding
```
I feel like this information would be better placed on a page describing everything there is to now about entries (bamboo.md?). This section could then _just_ describe what the endpoint is doing without also explaining how to encode entries.

adzialocha replied:
Hmn, I think it explains quite well _why_ this endpoint exists at all, not sure! I'll leave it here for now

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+
+type EntryArgsResponse {
+  """
+  log id to be used to forge the next entry
```
same as above

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+  """
+  CBOR representation of a signed Bamboo entry, encoded as a hexadecimal string
+  """
+  entryEncoded: String!
```
Let's call this `entry` and the other parameter `operation`!

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+
+- returns one document view of a given schema
     - no side effects
+- either a `id` or `view_id` needs to be specified
```
I moved this here because it doesn't apply for the `all_` query

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+  """
+  filter collection of documents
+  """
+  where: <schema_id>Filter
```
Should this be a list `[<schema_id>Filter]`?

adzialocha replied:
I think the syntax will be:

```
all_bla_0020...abc(where: { ... }) {
}
```

.. so that should be correct then!

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+    - documents can be queried by `document_view_id` in order to receive a [documents][view] onto it's data at a specific materialised state
 
-## Publishing Entries
+### Publishing Entries
```
The TOC currently doesn't display the actual operation names. We should have that eventually one way or the other! But doesn't need to be now.

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+
+  """
+  specific view id of the document to be queried, this returns the document
+  state at a certain point in time
```
I think it may be confusing when we say that a view id represents a point in time because we can never tell this point in time (especially for view ids with more than one operation id)

cafca commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,333 @@ id: queries
+
+#### `all_<schema_id>`
+
+- returns many document views of a given schema
```
I changed this to be more precise. Hope my change works for you :)

cafca commented:
It's good! Please check out the comments that I put eye emoji reactions on, I feel that these are kind of important.

adzialocha commented on website/docs/02-organising-data/queries.md:
```diff
@@ -4,49 +4,363 @@ id: queries
+    - the operation adheres to the [operation specification][operations]
+    - the operation is linked to the entry with a correct payload hash and size
+  - the operation may be materialised on the node resulting in a new document view
+- returns entry arguments required for publishing the next entry for the same document, similar to `nextEntryArgs`
```
That opens a little bit of a different topic for me, which is more like a new future specification related to node configuration? I would leave it in for now and we can see later how and what options of nodes we will offer?

adzialocha commented:
> It's good! Please check out the comments that I put eye emoji reactions on, I feel that these are kind of important.

Cool, thanks! I've did some changes but not all of them (even if they had cute eyes) as I'm not sure if that opens a too big discussion at that point as it feels slightly related to other concepts. Happy to move these things into issues though!
